### ISSUE TITLE
Simultaneous access to the same zarr chunk is browser-dependent

### ISSUE THREAD

**Issue by guigrpa** [gzuidhof/zarr.js#86]
Let's consider a program that accesses the same zarr chunk multiple times.

When the multiple requests are executed **serially**, and from what I've seen in the Chrome DevTools, the browser's cache is used to avoid fetching the chunk data multiple times. This works consistently across browsers:

```js
const z = await openArray({
  store: 'https://example.com',
  path: 'my-store.zarr',
  mode: 'r',
});
let data;
data = await fetchData(z, [0, slice(0, 5), slice(0, 5)]);
data = await fetchData(z, [0, slice(5, 10), slice(5, 10)]);
data = await fetchData(z, [0, slice(10, 15), slice(10, 15)]);

const fetchData = async (z, fragment) => {
  const res = await z.get(fragment);
  return res.data;
};
```

When requests are executed **in parallel, though**…

```js
data = await Promise.all([
  fetchData(z, [0, slice(0, 5), slice(0, 5)]),
  fetchData(z, [0, slice(5, 10), slice(5, 10)]),
  fetchData(z, [0, slice(10, 15), slice(10, 15)]),
]);
```

…behaviour depends on the browser:

* In Chrome and Firefox, the original request is duplicated, but no data is sent:
![Screenshot 2021-03-17 at 07 37 07](https://user-images.githubusercontent.com/2129481/111426816-82881d00-86f5-11eb-992e-4dce1c8a147d.png)
![Screenshot 2021-03-17 at 07 39 56](https://user-images.githubusercontent.com/2129481/111426830-86b43a80-86f5-11eb-8388-ea6e1f61925f.png)

* In Safari (and I guess Mobile Safari), though, the request is sent multiple times (if we look at the request & response headers, they have unique IDs, etc.):
![Screenshot 2021-03-17 at 07 45 28](https://user-images.githubusercontent.com/2129481/111426839-8a47c180-86f5-11eb-908c-50dd79fbfb63.png)

In our use case, preventing simultaneous access to (different types of) the same chunk is very hard. Would it be possible to cache chunks at the zarr.js library level to prevent multiple fetches? Or could you suggest a workaround?

Thanks for your state-of-the-art library!

**Comment by manzt** [comment]
Thanks for opening an issue, and I totally understand the concern. 

### What is a `Store`?

The `Store` interface in zarr.js, much like `zarr-python`, is intended to be very flexible and extended, so there are several ways you could get the desired behavior.

When you supply a url string to `store` like in the example above, internally `openArray` initializes an `HTTPStore` class:

https://github.com/gzuidhof/zarr.js/blob/b38fe8f1c35ccc3f8bf5ca91f10c9124de003ebc/src/storage/httpStore.ts#L18

The `ZarrArray` is agnostic to the storage backend so it doesn't make any assumptions about the overhead of calling `store.getItem`, e.g. whether it involves a network request. 

In your example, each time you call `z.get()`, since the slices are in the same chunk, `HTTPStore.getItem("0.0.0")` is called each time (which makes a fetch request). (It's worth noting that this is the behavior of most storage backends in zarr-python as well).

### Possible solution

A solution is to _compose_ an instance of `HTTPStore` with some type of cache that mimics that `Store` interface. This means that successive calls for the same key (`"0.0.0"`) will hit the cache.  However, we have intentionally kept store implementations minimal in this library (see #78 for more information). 

 In your project with zarr.js this could look something like:

```javascript
import { HTTPStore, openArray } from 'zarr';

class LRUStoreCache {
   constructor(store, cacheSize) {  /*... */ }
   getItem(key) {/* ... */} 
   containsItem(key) {/* ... */}
}

const store = new HTTPStore('https://example.com/my-store.zarr');

const z = await openArray({
  store: new LRUStoreCache(store, 10000),
  mode: 'r',
});

let data;
data = await fetchData(z, [0, slice(0, 5), slice(0, 5)]);
data = await fetchData(z, [0, slice(5, 10), slice(5, 10)]);
data = await fetchData(z, [0, slice(10, 15), slice(10, 15)]);

const fetchData = async (z, fragment) => {
  const res = await z.get(fragment);
  return res.data;
};
```

An example of `LRUStoreCache` can be found for `zarr-python` [here](https://zarr.readthedocs.io/en/stable/api/storage.html#zarr.storage.LRUStoreCache).

**Comment by guigrpa** [comment]
Thanks for the insights!

For the moment, I'm exploring an alternative route and caching the requests to the zarr library. If that's not enough, I'll get a shot at building the cached HTTPStore as you suggest.

**Comment by manzt** [comment]
If you are curious, I have implemented an `LRUCacheStore` using this approach for a project of mine: https://github.com/hms-dbmi/vizarr/blob/main/src/lru-store.ts

**Comment by guigrpa** [comment]
Thanks for sharing!

However, this implementation would still fetch N times the same block if it's requested simultaneously  N times by the zarr.js user, as in my original post:

```js
data = await Promise.all([
  fetchData(z, [0, slice(0, 5), slice(0, 5)]),
  fetchData(z, [0, slice(5, 10), slice(5, 10)]),
  fetchData(z, [0, slice(10, 15), slice(10, 15)]),
]);
```

In this case, I believe the `LRUCacheStore` should cache promised data, rather than the actual data.

**Comment by gzuidhof** [comment]
Just chiming in here: if the LRU cache instead contained promises of the data then indeed it would solve this behavior, just make sure to delete it from the cache if the promise rejects (and possibly after a timeout)! Otherwise if you have a spotty internet connection / failing server there is no way to retry.

Probably out of scope but another possible improvement *weighing* the cache items (for lack of a better name). So one could specify that they want the cache to be 1000MB, and after inserting one would evict items until they are under the threshold again (i.e. the size is defined in bytes instead of items). I once adapted a package to do that.. but sadly that code belongs to my former employer :(.

**Comment by esiegel** [comment]
The caching strategy above definitely works, but it still has the additional cost of decoding the chunked data multiple times.  If you are requesting the same data many times this can become a big issue as I have seen.

Do you have thoughts on how to patch `getSelection` to potentially have something that caches the processed chunk.
